File: src/Core/Bindings/vili/vili.cpp
#include <Bindings/vili/Vili.hpp>

#include <System/Path.hpp>
#include <vili/config.hpp>
#include <vili/node.hpp>
#include <vili/types.hpp>

#include <Bindings/Config.hpp>

namespace vili::bindings
{
    void load_enum_node_type(sol::state_view state)
    {
        sol::table vili_namespace = state["vili"].get<sol::table>();
        vili_namespace.new_enum<vili::node_type>("node_type",
            { { "null", vili::node_type::null }, { "string", vili::node_type::string },
                { "integer", vili::node_type::integer }, { "number", vili::node_type::number },
                { "boolean", vili::node_type::boolean }, { "array", vili::node_type::array },
                { "object", vili::node_type::object } });
    }
    void load_class_const_node_iterator(sol::state_view state)
    {
        sol::table vili_namespace = state["vili"].get<sol::table>();
        sol::usertype<vili::const_node_iterator> bind_const_node_iterator
            = vili_namespace.new_usertype<vili::const_node_iterator>("const_node_iterator",
                sol::call_constructor,
                sol::constructors<vili::const_node_iterator(array::const_iterator),
                    vili::const_node_iterator(object::const_iterator),
                    vili::const_node_iterator(const vili::const_node_iterator&)>());
        bind_const_node_iterator["operator++"] = &vili::const_node_iterator::operator++;
        bind_const_node_iterator[sol::meta_function::multiplication]
            = &vili::const_node_iterator::operator*;
    }
    void load_class_node(sol::state_view state)
    {
        sol::table vili_namespace = state["vili"].get<sol::table>();
        sol::usertype<vili::node> bind_node
            = vili_namespace.new_usertype<vili::node>("node", sol::call_constructor,
                sol::constructors<vili::node(), vili::node(int), vili::node(vili::integer),
                    vili::node(vili::number), vili::node(const vili::string&),
                    vili::node(std::string_view), vili::node(vili::boolean),
                    vili::node(const char*), vili::node(const vili::array&),
                    vili::node(const vili::object&), vili::node(const vili::node&)>());
        bind_node["type"] = &vili::node::type;
        bind_node["dump"]
            = sol::overload([](vili::node* self) -> std::string { return self->dump(); },
                [](vili::node* self, bool root) -> std::string { return self->dump(root); });
        bind_node["is_primitive"] = &vili::node::is_primitive;
        bind_node["is_container"] = &vili::node::is_container;
        bind_node["is_null"] = &vili::node::is_null;
        bind_node["is_integer"] = &vili::node::is_integer;
        bind_node["is_number"] = &vili::node::is_number;
        bind_node["is_numeric"] = &vili::node::is_numeric;
        bind_node["is_boolean"] = &vili::node::is_boolean;
        bind_node["is_string"] = &vili::node::is_string;
        bind_node["is_array"] = &vili::node::is_array;
        bind_node["is_object"] = &vili::node::is_object;
        bind_node["as_boolean"] = &vili::node::as_boolean;
        bind_node["as_integer"] = &vili::node::as_integer;
        bind_node["as_number"] = &vili::node::as_number;
        bind_node["as_string"] = &vili::node::as_string;
        bind_node["as_array"] = &vili::node::as_array;
        bind_node["as_object"] = &vili::node::as_object;
        bind_node[sol::meta_function::index] = sol::overload(
            static_cast<vili::node& (vili::node::*)(const char*)>(&vili::node::operator[]),
            static_cast<vili::node& (vili::node::*)(const std::string&)>(&vili::node::operator[]),
            static_cast<vili::node& (vili::node::*)(size_t)>(&vili::node::operator[]),
            static_cast<vili::node& (vili::node::*)(unsigned int)>(&vili::node::operator[]),
            static_cast<const vili::node& (vili::node::*)(const char*) const>(
                &vili::node::operator[]),
            static_cast<const vili::node& (vili::node::*)(const std::string&) const>(
                &vili::node::operator[]),
            static_cast<const vili::node& (vili::node::*)(size_t) const>(&vili::node::operator[]));
        bind_node["push"] = &vili::node::push;
        bind_node["insert"] = sol::overload(
            static_cast<void (vili::node::*)(size_t, const vili::node&)>(&vili::node::insert),
            static_cast<void (vili::node::*)(const std::string&, vili::node)>(&vili::node::insert));
        bind_node["merge"] = &vili::node::merge;
        bind_node["contains"] = &vili::node::contains;
        bind_node["erase"]
            = sol::overload(static_cast<void (vili::node::*)(size_t)>(&vili::node::erase),
                static_cast<void (vili::node::*)(size_t, size_t)>(&vili::node::erase),
                static_cast<void (vili::node::*)(const std::string&)>(&vili::node::erase));
        bind_node["front"] = &vili::node::front;
        bind_node["back"] = &vili::node::back;
        bind_node["begin"]
            = sol::overload(static_cast<vili::node_iterator (vili::node::*)()>(&vili::node::begin),
                static_cast<vili::const_node_iterator (vili::node::*)() const>(&vili::node::begin));
        bind_node["end"]
            = sol::overload(static_cast<vili::node_iterator (vili::node::*)()>(&vili::node::end),
                static_cast<vili::const_node_iterator (vili::node::*)() const>(&vili::node::end));
        bind_node["items"]
            = sol::overload(static_cast<vili::object& (vili::node::*)()>(&vili::node::items),
                static_cast<const vili::object& (vili::node::*)() const>(&vili::node::items));
        bind_node["at"] = sol::overload(
            static_cast<vili::node& (vili::node::*)(const std::string&)>(&vili::node::at),
            static_cast<vili::node& (vili::node::*)(size_t)>(&vili::node::at),
            static_cast<const vili::node& (vili::node::*)(const std::string&) const>(
                &vili::node::at),
            static_cast<const vili::node& (vili::node::*)(size_t) const>(&vili::node::at));
        bind_node["data"] = &vili::node::data;
        bind_node["size"] = &vili::node::size;
        bind_node["empty"] = &vili::node::empty;
        bind_node["clear"] = &vili::node::clear;
        bind_node[sol::meta_function::equal_to] = &vili::node::operator==;
        bind_node["from_type"] = &vili::node::from_type;
        state.script_file("obe://Lib/Internal/Vili.lua"_fs);
    }
    void load_class_node_iterator(sol::state_view state)
    {
        sol::table vili_namespace = state["vili"].get<sol::table>();
        sol::usertype<vili::node_iterator> bind_node_iterator
            = vili_namespace.new_usertype<vili::node_iterator>("node_iterator",
                sol::call_constructor,
                sol::constructors<vili::node_iterator(array::iterator),
                    vili::node_iterator(object::iterator),
                    vili::node_iterator(const vili::node_iterator&)>());
        bind_node_iterator["operator++"] = &vili::node_iterator::operator++;
        bind_node_iterator[sol::meta_function::multiplication] = &vili::node_iterator::operator*;
    }
    void load_function_from_string(sol::state_view state)
    {
        sol::table vili_namespace = state["vili"].get<sol::table>();
        vili_namespace.set_function("from_string", &vili::from_string);
    }
    void load_function_to_string(sol::state_view state)
    {
        sol::table vili_namespace = state["vili"].get<sol::table>();
        vili_namespace.set_function("to_string", &vili::to_string);
    }
    void load_global_permissive_cast(sol::state_view state)
    {
        sol::table vili_namespace = state["vili"].get<sol::table>();
        vili_namespace["PERMISSIVE_CAST"] = vili::PERMISSIVE_CAST;
    }
    void load_global_verbose_exceptions(sol::state_view state)
    {
        sol::table vili_namespace = state["vili"].get<sol::table>();
        vili_namespace["VERBOSE_EXCEPTIONS"] = vili::VERBOSE_EXCEPTIONS;
    }
    void load_global_true_value(sol::state_view state)
    {
        sol::table vili_namespace = state["vili"].get<sol::table>();
        vili_namespace["true_value"] = vili::true_value;
    }
    void load_global_false_value(sol::state_view state)
    {
        sol::table vili_namespace = state["vili"].get<sol::table>();
        vili_namespace["false_value"] = vili::false_value;
    }
    void load_global_null_typename(sol::state_view state)
    {
        sol::table vili_namespace = state["vili"].get<sol::table>();
        vili_namespace["null_typename"] = vili::null_typename;
    }
    void load_global_boolean_typename(sol::state_view state)
    {
        sol::table vili_namespace = state["vili"].get<sol::table>();
        vili_namespace["boolean_typename"] = vili::boolean_typename;
    }
    void load_global_integer_typename(sol::state_view state)
    {
        sol::table vili_namespace = state["vili"].get<sol::table>();
        vili_namespace["integer_typename"] = vili::integer_typename;
    }
    void load_global_number_typename(sol::state_view state)
    {
        sol::table vili_namespace = state["vili"].get<sol::table>();
        vili_namespace["number_typename"] = vili::number_typename;
    }
    void load_global_string_typename(sol::state_view state)
    {
        sol::table vili_namespace = state["vili"].get<sol::table>();
        vili_namespace["string_typename"] = vili::string_typename;
    }
    void load_global_object_typename(sol::state_view state)
    {
        sol::table vili_namespace = state["vili"].get<sol::table>();
        vili_namespace["object_typename"] = vili::object_typename;
    }
    void load_global_array_typename(sol::state_view state)
    {
        sol::table vili_namespace = state["vili"].get<sol::table>();
        vili_namespace["array_typename"] = vili::array_typename;
    }
    void load_global_unknown_typename(sol::state_view state)
    {
        sol::table vili_namespace = state["vili"].get<sol::table>();
        vili_namespace["unknown_typename"] = vili::unknown_typename;
    }
    void load_global_container_typename(sol::state_view state)
    {
        sol::table vili_namespace = state["vili"].get<sol::table>();
        vili_namespace["container_typename"] = vili::container_typename;
    }
};

File: engine/Hints/vili.lua
---@meta

vili = {};
---@class vili.const_node_iterator
vili._const_node_iterator = {};

--- vili.const_node_iterator constructor
---
---@param value array.const_iterator #
---@return vili.const_node_iterator
function vili.const_node_iterator(value) end

--- vili.const_node_iterator constructor
---
---@param value object.const_iterator #
---@return vili.const_node_iterator
function vili.const_node_iterator(value) end

--- vili.const_node_iterator constructor
---
---@param other_it vili.const_node_iterator #
---@return vili.const_node_iterator
function vili.const_node_iterator(other_it) end



---@class vili.node
vili._node = {};

--- Default constructor, node will have null type.
---
---@return vili.node
function vili.node() end

--- Same than vili::integer constructor.
---
---@param value number #
---@return vili.node
function vili.node(value) end

--- Creates a node that contains an integer.
---
---@param value vili.integer #
---@return vili.node
function vili.node(value) end

--- Creates a node that contains a number.
---
---@param value vili.number #
---@return vili.node
function vili.node(value) end

--- Creates a node that contains a string.
---
---@param value vili.string #
---@return vili.node
function vili.node(value) end

--- Creates a node that contains a string.
---
---@param value string #
---@return vili.node
function vili.node(value) end

--- Creates a node that contains a boolean.
---
---@param value vili.boolean #
---@return vili.node
function vili.node(value) end

--- Same than vili::string constructor.
---
---@param value string #
---@return vili.node
function vili.node(value) end

--- Creates a node that contains an array (vector-like container)
---
---@param value vili.array #
---@return vili.node
function vili.node(value) end

--- Creates a node that contains a object (map-like container)
---
---@param value vili.object #
---@return vili.node
function vili.node(value) end

--- node copy constructor
---
---@param copy vili.node #
---@return vili.node
function vili.node(copy) end

--- node move constructor
---
---@param move vili.node #
---@return vili.node
function vili.node(move) end


--- Retrieves the type of the underlying value of the node.
---
---@return vili.node_type
function vili._node:type() end

--- Dumps the node content as a string.
---
---@param root? boolean #
---@return string
function vili._node:dump(root) end

--- Checks if the node contains a given type.
---
---@return boolean
function vili._node:is() end

--- Checks if the node contains a given type.
---
---@return boolean
function vili._node:is() end

--- Checks whether the underlying value is a primitive (boolean, integer, number, string) or not.
---
---@return boolean
function vili._node:is_primitive() end

--- Checks whether the underlying value is a container (array, object) or not.
---
---@return boolean
function vili._node:is_container() end

--- Checks whether the node contains a value or node.
---
---@return boolean
function vili._node:is_null() end

--- Checks whether the underlying value is an integer.
---
---@return boolean
function vili._node:is_integer() end

--- Checks whether the underlying value is a number.
---
---@return boolean
function vili._node:is_number() end

--- Checks whether the underlying value is an integer or a number.
---
---@return boolean
function vili._node:is_numeric() end

--- Checks whether the underlying value is a boolean.
---
---@return boolean
function vili._node:is_boolean() end

--- Checks whether the underlying value is a string.
---
---@return boolean
function vili._node:is_string() end

--- Checks whether the underlying value is an array.
---
---@return boolean
function vili._node:is_array() end

--- Checks whether the underlying value is an object.
---
---@return boolean
function vili._node:is_object() end

--- Returns the node as the underlying type.
---
---@return T
function vili._node:as() end

--- Returns the node as the underlying type.
---
---@return std.enable_if_t[!std.is_floating_point_v[T >||! vili.PERMISSIVE_CAST,T]]
function vili._node:as() end

--- Returns the node as the underlying type This overload is only used when vili has to convert an integer to a floating point number and therefore can't return a const reference from the internal variant.
---
---@return std.enable_if_t[std.is_floating_point_v[T > && vili.PERMISSIVE_CAST,T]]
function vili._node:as() end

---@return std.enable_if_t[!std.is_floating_point_v[T >||! vili.PERMISSIVE_CAST,T]]
function vili._node:as() end

--- Returns the node as a boolean.
---
---@return vili.boolean
function vili._node:as_boolean() end

--- Returns the node as an integer.
---
---@return vili.integer
function vili._node:as_integer() end

--- Returns the node as a number.
---
---@return vili.number
function vili._node:as_number() end

--- Returns the node as a string.
---
---@return vili.string
function vili._node:as_string() end

--- Returns the node as an array.
---
---@return vili.array
function vili._node:as_array() end

--- Returns the node as an object.
---
---@return vili.object
function vili._node:as_object() end

---@param value vili.node #
function vili._node:push(value) end

--- Emplace a child node at given index.
---
---@param index number #Index of the node to emplace
---@param value value_type #Value of the node to emplace
function vili._node:emplace(index, value) end

--- Emplace a child node at given key.
---
---@param key string #Key of the node to emplace
---@param value value_type #Value of the node to emplace
function vili._node:emplace(key, value) end

---@param index number #
---@param value vili.node #
function vili._node:insert(index, value) end

---@param key string #
---@param value vili.node #
function vili._node:insert(key, value) end

---@param value vili.node #
function vili._node:merge(value) end

---@param key string #
---@return boolean
function vili._node:contains(key) end

---@param index number #
function vili._node:erase(index) end

---@param begin number #
---@param end number #
function vili._node:erase(begin, _end) end

---@param key string #
function vili._node:erase(key) end

---@return vili.node
function vili._node:front() end

---@return vili.node
function vili._node:back() end

---@return vili.node_iterator
function vili._node:begin() end

---@return vili.const_node_iterator
function vili._node:begin() end

---@return vili.node_iterator
function vili._node:_end() end

---@return vili.const_node_iterator
function vili._node:_end() end

---@return vili.object
function vili._node:items() end

---@return vili.object
function vili._node:items() end

---@param key string #
---@return vili.node
function vili._node:at(key) end

---@param index number #
---@return vili.node
function vili._node:at(index) end

---@param key string #
---@return vili.node
function vili._node:at(key) end

---@param index number #
---@return vili.node
function vili._node:at(index) end

--- Directly access underlying variant.
---
---@return vili.node_data
function vili._node:data() end

---@return number
function vili._node:size() end

---@return boolean
function vili._node:empty() end

function vili._node:clear() end

---@param type vili.node_type #
---@return vili.node
function vili._node:from_type(type) end


---@class vili.node_iterator
vili._node_iterator = {};

--- vili.node_iterator constructor
---
---@param value array.iterator #
---@return vili.node_iterator
function vili.node_iterator(value) end

--- vili.node_iterator constructor
---
---@param value object.iterator #
---@return vili.node_iterator
function vili.node_iterator(value) end

--- vili.node_iterator constructor
---
---@param other_it vili.node_iterator #
---@return vili.node_iterator
function vili.node_iterator(other_it) end




---@alias vili.node_data nil|vili.object|vili.array|vili.integer|vili.number|vili.boolean|vili.string

---@alias vili.null nil

---@alias vili.object nlohmann.fifo_map[string,vili.node]

---@alias vili.array vili.node[]

---@alias vili.integer number

---@alias vili.number number

---@alias vili.boolean boolean

---@alias vili.string string
---@param type string #
---@return vili.node_type
function vili.from_string(type) end

---@param type vili.node_type #
---@return string
function vili.to_string(type) end


---@type boolean
vili.PERMISSIVE_CAST = {};

---@type boolean
vili.VERBOSE_EXCEPTIONS = {};

---@type string
vili.true_value = {};

---@type string
vili.false_value = {};

---@type string
vili.null_typename = {};

---@type string
vili.boolean_typename = {};

---@type string
vili.integer_typename = {};

---@type string
vili.number_typename = {};

---@type string
vili.string_typename = {};

---@type string
vili.object_typename = {};

---@type string
vili.array_typename = {};

---@type string
vili.unknown_typename = {};

---@type string
vili.container_typename = {};

--- An enum representing Types of existing Nodes.
---
---@class vili.node_type
vili.node_type = {
    ---@type vili.node_type
    null = 0,
    ---@type vili.node_type
    string = 1,
    ---@type vili.node_type
    integer = 2,
    ---@type vili.node_type
    number = 3,
    ---@type vili.node_type
    boolean = 4,
    ---@type vili.node_type
    array = 5,
    ---@type vili.node_type
    object = 6,
};
return vili;